C/deserialize.h

```diff
@@ -69,23 +38,30 @@ int32_t decodeUptoMaxInt(bit_stream* stream);
  *                          of the function is positive and when NULL != census;
  *                NULL == *dag when the return value is negative.
  */
-int32_t decodeMallocDag(dag_node** dag, combinator_counters* census, bit_stream* stream);
+int32_t decodeMallocDag(dag_node** dag, combinator_counters* census, bitstream* stream);
 
 /* Decode a string of up to 2^31 - 1 bits from 'stream'.
  * This is the format in which the data for 'WITNESS' nodes are encoded.
  * Returns 'ERR_DATA_OUT_OF_RANGE' if the encoded string of bits exceeds this decoder's limits.
  * Returns 'ERR_BITSTRING_EOF' if not enough bits are available in the 'stream'.
  * Returns 'ERR_BITSTREAM_ERROR' if an I/O error occurs when reading from the 'stream'.
  * Returns 'ERR_MALLOC' if malloc fails.
- * If successful, '*witness' is set to the decoded bitstring and 0 is returned.
+ * If successful, '*witness' is set to the decoded bitstring,
+ *                '*allocation' points to memory allocated for this bitstring,
+ *                and 0 is returned.
+ *
+ * If an error is returned '*allocation' is set to 'NULL' and '*witness' might be modified.
  *
- * Precondition: NULL != witness;
+ * Precondition: NULL != allocation;
+ *               NULL != witness;
+ *               NULL != stream;
  *
- * Postcondition: if 'result == 0' and '0 < witness->len' then
- *                  'unsigned char witness->arr[(witness->len + witness->offset - 1) / CHAR_BIT + 1] and
- *                  (*witness) represents a some bitstring;
- *                if 'result < 0' or 'wintess->len == 0' then witness->arr == NULL;
+ * Postcondition: if the function returns '0'
+ *                  '*witness' represents a some bitstring
+ *                   and '*allocation' points to allocated memory sufficient for at least the 'witness->arr' array.
+ *                       (Note that '*allocation' may be 'NULL' if 'n == 0')
+ *                if the function returns a negative value then '*allocation == NULL'
  */
-int32_t decodeMallocWitnessData(bitstring* witness, bit_stream* stream);
+int32_t decodeMallocWitnessData(void** allocation, bitstring* witness, bitstream* stream);
```

Comment:
I think it'd be a cleaner API if there were a function that took a `bitstring*` and freed its memory, rather than returning the allocation through an outptr and requiring the user free it themselves.

Reply:
I did try that design but there is a major problem in that some `bitstring`s own their memory and (most) other `bitstring`s share their memory.  The solution means that `bitstrings` all share their memory and someone else owns it.

There is also a minor problem that `bistring`s have pointers to `const` and thus cannot be freed.  While not a major problem by itself, it does guide the solution in this direction.